### PR TITLE
Add checkpoints

### PR DESCRIPTION
**Pull request by ravwojdyla**

### CODE REVIEW

**Comment by yonromai** [comment]
Looks really useful! I've played with checkpoints on my own and came up with a very similar API 
* Was considering pimping SCollection to avoid have to pass sc as a parameter
* Also, wondering if we could avoid having to name the checkpoint by using some hash of the function/context, probably over-engineering though 
 * An example would be useful I think, although tests might be enough

**Comment by yonromai** [comment]
oh, well, hopefully it makes it into beam :)

**Comment by ravwojdyla** [comment]
@nevillelyh @andrewsmartin any comments?

File: scio-core/src/main/scala/com/spotify/scio/util/ScioUtil.scala [hunk 1]
```diff
@@ -84,4 +89,32 @@ private[scio] object ScioUtil {
       s"Could not get dataflow metrics of ${getMetrics.getJobId} in ${getMetrics.getProjectId}")
   }
 
+  def addPartSuffix(path: String, ext: String = ""): String =
+    if (path.endsWith("/")) s"${path}part-*$ext" else s"$path/part-*$ext"
+
+  def getTempFile(@transient context: ScioContext, fileOrPath: String = null): String = {
```
Comment (andrewsmartin) [inline scio-core/src/main/scala/com/spotify/scio/util/ScioUtil.scala]: What does the `@transient` do here? I've only seen it used for fields.

File: scio-extra/src/main/scala/com/spotify/scio/extra/checkpoint/package.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+     * Checkpoints are useful for debugging one part of a long flow, when you would otherwise have
+     * to run many steps to get to the one you care about.  To enable checkpoints, sprinkle calls to
+     * [[checkpoint[T]()]] throughout your flow, ideally after expensive steps.
+     *
+     * @param fileOrPath filename or fully qualified path to the checkpoint
+     * @param fn result of this arbitrary [[ScioContext]] => [[SCollection[T]] flow is what is
+     *           checkpointed
+     */
+    def checkpoint[T: ClassTag](fileOrPath: String,
+                                fn: ScioContext => SCollection[T]): SCollection[T] = {
```
Comment (andrewsmartin) [inline scio-extra/src/main/scala/com/spotify/scio/extra/checkpoint/package.scala]: Would it be good to have the `fn` in a separate parameter list? That way blocks could be used.

**Comment by andrewsmartin** [comment]
Does this work for specifying multiple checkpoints?

File: scio-core/src/main/scala/com/spotify/scio/util/ScioUtil.scala [hunk 3]
```diff
@@ -18,23 +18,28 @@
...
+import org.apache.beam.sdk.options.{GcpOptions, PipelineOptions}
 import org.apache.beam.sdk.util.Transport
+import org.slf4j.LoggerFactory
 
 import scala.reflect.ClassTag
+import scala.util.{Failure, Success}
 
 private[scio] object ScioUtil {
 
+  @transient lazy private val log = LoggerFactory.getLogger(this.getClass)
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/util/ScioUtil.scala]: `private lazy val` so it's easy to grep for `lazy val`?
~There's one in `SCollection` as well but we can fix that separately.~ Just realized this PR removed that one.

File: scio-core/src/main/scala/com/spotify/scio/util/ScioUtil.scala [hunk 4]
```diff
@@ -84,4 +89,32 @@ private[scio] object ScioUtil {
...
+    if ((ScioUtil.isLocalUri(uri) && uri.toString.startsWith("/")) || uri.isAbsolute) {
+      fop
+    } else {
+      val filename = fop
+      val tmpDir = if (context.options.getTempLocation != null) {
+        context.options.getTempLocation
+      } else {
+        val m =
+          "Specify a temporary location via --tempLocation or PipelineOptions.setTempLocation."
+        scala.util.Try(context.optionsAs[GcpOptions].getGcpTempLocation) match {
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/util/ScioUtil.scala]: Import `Try` as well? It's commonly used and not need to use full name.

**Comment by ravwojdyla** [comment]
@andrewsmartin 
 * https://issues.apache.org/jira/browse/BEAM-1458
 * should work for multiple checkpoints - added it as part of a job test

**Comment by ravwojdyla** [comment]
@andrewsmartin @nevillelyh updated to code:
  * made it easier to test job with checkpoints
  * update tempLocation before pipeline creation

Any other comments?

File: scio-examples/src/main/scala/com/spotify/scio/examples/extra/CheckpointExample.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+package com.spotify.scio.examples.extra
+
+import com.spotify.scio.ContextAndArgs
+import com.spotify.scio.extra.checkpoint._
+
+object CheckpointExample {
+  def main(cmdlineArgs: Array[String]): Unit = {
+    val (sc, args) = ContextAndArgs(cmdlineArgs)
+
+    val words = sc.checkpoint(args("checkpoint") + "-src"){
```
Comment (nevillelyh) [inline scio-examples/src/main/scala/com/spotify/scio/examples/extra/CheckpointExample.scala]: Add a few comments here explaining what they do? Also space between `) {`

File: scio-examples/src/main/scala/com/spotify/scio/examples/extra/CheckpointExample.scala [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      sc.textFile(args("input"))
+        .map(_.trim)
+        .flatMap(_.split("[^a-zA-Z']+").filter(_.nonEmpty))
+      }
```
Comment (nevillelyh) [inline scio-examples/src/main/scala/com/spotify/scio/examples/extra/CheckpointExample.scala]: `}` should align with `val words` not `sc.textFile`?

File: scio-extra/src/main/scala/com/spotify/scio/extra/checkpoint/package.scala [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ * import com.spotify.scio.extra.checkpoint._
+ * }}}
+ */
+package object checkpoint {
+
+  // scalastyle:off method.name
+  /**
+   * For use in testing, see [[CheckpointExampleTest]].
+   */
+  def CheckpointIO[T](fileOrPath: String): ObjectFileIO[T] = ObjectFileIO[T](fileOrPath)
```
Comment (nevillelyh) [inline scio-extra/src/main/scala/com/spotify/scio/extra/checkpoint/package.scala]: This is cool. I wonder if it's worth changing some other TestIOs to use this trick? IIRC `ProtoIO` also uses `ObjectFileIO` under the hood?

Comment (ravwojdyla) [inline scio-extra/src/main/scala/com/spotify/scio/extra/checkpoint/package.scala]: Would probably not buy us much given that there are two paths for saveAsPB - one for testing with protobuf objects and other after conversion to avro. In this case checkpoints use object IO for both paths, so had to do this trick.

**Pull request merged**